README.md
# multielo

This package implements a multiplayer extension of the popular Elo rating
system.

 * [Installation](#installation)
 * [Example Usage](#example-usage)
 * [Methodology](#methodology)
   * [Traditional Elo ratings](#traditional-elo-ratings)
   * [Extension to multiplayer](#extension-to-multiplayer)

## Installation

The package can be installed from GitHub with using ``pip``.

```bash
pip install git+git://github.com/djcunningham0/multielo.git
```

You can install a specific released version of the package using tag names. For
example, to install release v0.1.0 you can use:

```bash
pip install git+git://github.com/djcunningham0/multielo.git@v0.1.0
```

## Example Usage

The following example shows how to calculate updated Elo ratings after a 
matchup using the default settings in the package.

```python
>>> from multielo import MultiElo
>>> import numpy as np
>>>
>>> elo = MultiElo()
>>>
>>> # player with 1200 rating beats a player with 1000 rating
>>> elo.get_new_ratings(np.array([1200, 1000]))
array([1207.68809835,  992.31190165])
>>>
>>> # player with 900 rating beats player with 1000 rating
>>> elo.get_new_ratings(np.array([900, 1000]))
array([920.48207999, 979.51792001])
>>>
>>> # 3-way matchup
>>> elo.get_new_ratings(np.array([1200, 900, 1000]))
array([1208.34629612,  910.43382278,  981.21988111])
```

See [`demo.ipynb`](https://github.com/djcunningham0/multielo/blob/master/demo.ipynb) 
for a more in-depth tutorial, including details on parameters that can be tuned
in the Elo algorithm and examples of how to use the Tracker and Player objects 
to keep track of Elo ratings for a group of players over time.

## Methodology

*Note: This multiplayer extension isn't very mathematically or statistically
rigorous. It seems to work reasonably well, but I don't think the rating values
are as interpretable as traditional Elo. Send me a note or open an issue if you
have ideas for improving the methodology.*

### Traditional Elo ratings

Traditional Elo ratings only apply to 1-on-1 matchups. The new ratings for the
players involved in a matchup are calculated from:
1. The initial ratings before the matchup.
2. The matchup result.

There are also a few parameters that must be set for an Elo rating system:
1. *K* controls how many Elo rating points are gained or lost in a single game.
Larger *K* results in larger changes after each matchup. A commonly used value
is *K* = 32.
2. *D* controls the estimated win probability of each player. A commonly used 
value is *D* = 400, which means that a player with a 200-point Elo advantage is
expected to win ~75% of the time. A smaller *D* value means the player with the
higher Elo rating has a higher estimated win probability.

Suppose we have two players, *A* and *B*, with initial ratings *R<sub>A</sub>* 
and *R<sub>B</sub>*, respectively. We calculate an "expected score" for each 
player according to those ratings. The expected score for player *A* is:

![equation](https://latex.codecogs.com/png.latex?E_A%20%3D%20%5Cfrac%7B1%7D%7B1%20&plus;%2010%5E%7B%28R_B%20-%20R_A%29%20/%20D%7D%7D)

and the expected score for player *B* is defined similarly. 

To calculate the players' new ratings, we compare these expected scores with 
the "actual scores", or the matchup result. The actual score for player *A* is:

![equation](https://latex.codecogs.com/png.latex?S_A%20%3D%20%5Cleft%5C%7B%5Cbegin%7Bmatrix%7D%201%20%26%20%5Ctext%7Bplayer%20%5Ctextit%7BA%7D%20wins%7D%20%5C%5C%200%20%26%20%5Ctext%7Bplayer%20%5Ctextit%7BA%7D%20loses%7D%20%5Cend%7Bmatrix%7D%5Cright.)

Then the new rating for player *A*, *R'<sub>A</sub>*, is calculated from the
initial rating, expected score, and actual score:

![equation](https://latex.codecogs.com/png.latex?R%27_A%20%3D%20R_A&plus;K%28S_A%20-%20E_A%29)

### Extension to multiplayer

At minimum, a reasonable extension of Elo to multiplayer matchups should have 
the following properties:
1. Changes in ratings should be zero sum, just like traditional Elo.
2. Coming in first place is better (improves your rating more) than coming in
second place, is better than third place, and so on. Losing (coming in last 
place) cannot improve your rating.
3. It should converge to traditional Elo when there are only two players in the
matchup.

**Expected scores**

The first issue we must address is that the Elo expected score calculation only
involves the ratings of two players. In order to generalize this calculation to
multiplayer matchups, we can treat a single multiplayer matchup as a 
combination set of all pairwise 1-on-1 matchups between every player involved. 
Specifically, in a multiplayer matchup with *N* players there would be

![equation](https://latex.codecogs.com/png.latex?%5Cfrac%7BN%28N-1%29%7D%7B2%7D)

pairwise 1-on-1 matchups between the players.

To calculate each player's expected score for the multiplayer matchup, first we
calculate that player's expected score for each individual matchup just as with
traditional Elo. Then we scale the scores by *N(N-1)/2* so that the expected 
scores for all players sum to 1 (the expected scores are meant to represent 
probabilities so they must sum to 1).

The full expression for the expected score for player *A* in a multiplayer 
matchup with *N* players is:

![equation](https://latex.codecogs.com/png.latex?E_A%20%3D%20%5Cfrac%7B%5Csum_%7B1%20%5Cle%20i%20%5Cle%20N%2C%20i%20%5Cne%20A%7D%20%5Cfrac%7B1%7D%7B1%20&plus;%2010%5E%7B%28R_i%20-%20R_A%29/D%7D%7D%7D%7B%5Cfrac%7BN%28N-1%29%7D%7B2%7D%7D)

**Actual scores**

The "actual scores" in traditional are binary: 1 point for a win, 0 for a loss.
In a multiplayer matchup there are more than two options, so we need a 
different definition for actual scores. We can introduce the idea of a "score
function", where the actual score for player *A* is given by:

![equation](https://latex.codecogs.com/png.latex?S_A%20%3D%20%5Cleft%5C%7B%5Cbegin%7Bmatrix%7D%20s_1%20%26%20%5Ctext%7Bplayer%20%5Ctextit%7BA%7D%20finishes%201st%20out%20of%20%5Ctextit%7Bn%7D%7D%20%5C%5C%20s_2%20%26%20%5Ctext%7Bplayer%20%5Ctextit%7BA%7D%20finishes%202nd%20out%20of%20%5Ctextit%7Bn%7D%7D%20%5C%5C%20%5Cdots%20%26%20%5Cdots%20%5C%5C%20s_n%20%26%20%5Ctext%7Bplayer%20%5Ctextit%7BA%7D%20finishes%20%5Ctextit%7Bn%7Dth%20out%20of%20%5Ctextit%7Bn%7D%7D%20%5Cend%7Bmatrix%7D%5Cright.)

This scoring function should have a few properties:
1. The scores must sum to 1 to match the expected scores.
2. The scores should be monotonically decreasing. That is, 1st place should 
have a higher score than 2nd, which has a higher score than 3rd, and so on.
3. Last place should have a score of 0 because a last-place finish should
never improve a player's rating.

Note that when there are only two players, the only score function that
satisfies all three conditions is the binary definition.

With more than two players, there are many possible score functions that
satisfy all three conditions. Perhaps the most natural score function is what 
is defined in this implementation as a *linear score function*:

![equation](https://latex.codecogs.com/png.latex?S_A%5E%7Blinear%7D%20%3D%20%5Cfrac%7BN%20-%20p_A%7D%7BN%28N-1%29/2%7D)

where *N* is the number of players and *p<sub>A</sub>* is the place that player
*A* finishes in (1 for first place, 2 for second, and so on). For example, with
five players, the actual scores awarded to the players in order of finish are 
[0.4, 0.3, 0.2, 0.1, 0]. When using the linear score function, each improvement
in place is weighted equally. That is, improving from second to first has the 
same impact to the player's actual score as improving from third to second, and
so on.

In some cases it may be of interest to give more value to the top finishers. 
For example, in a poker game where only the top three finishers are paid out,
it may be optimal for players to take risks that improve their chances of
coming in first place rather than minimizing their chance of finishing in last.
That is, players will try to finish at the top, but there is very little 
real-world difference in finishing in last rather than second-to-last. One
solution to these cases is to use what is defined in this implementation as an
*exponential score function*:

![equation](https://latex.codecogs.com/png.latex?S_A%5E%7Bexp%7D%20%3D%20%5Cfrac%7B%5Calpha%5E%7BN-p_A%7D%20-%201%7D%7B%5Csum_%7Bi%3D1%7D%5E%7BN%7D%28%5Calpha%5E%7BN-i%7D%20-%201%29%7D)

where *N* is the number of players, *p<sub>A</sub>* is the place that player 
*A* finishes in, and ![equation](https://latex.codecogs.com/png.latex?%5Calpha) 
is the "base" of the score function. When ![equation](https://latex.codecogs.com/png.latex?%5Calpha%20%3E%201),
this score function places extra weight on finishing in the top places. For
example, when ![equation](https://latex.codecogs.com/png.latex?%5Calpha%20%3D%201.5)
and there are five players, the actual scores awarded to the players in order 
of finish are approximately [0.496, 0.290, 0.153, 0.061, 0]. A larger 
![equation](https://latex.codecogs.com/png.latex?%5Calpha) value increases the 
benefit to finishing in the top places. 

The chart below shows how three different score functions would assign actual 
scores in a 5-player matchup.

![image](images/score_function_plot.png)

It is also worth noting that as ![equation](https://latex.codecogs.com/png.latex?%5Calpha) 
approaches 1, the exponential score function converges to the linear score
function.

The ![equation](https://latex.codecogs.com/png.latex?%5Calpha) value in the 
exponential score function corresponds to the score_function_base parameter of
the `MultiElo` object in this package.

**Calculating new ratings**

The new rating calculation is very similar to traditional Elo. We compare the
actual and expected scores for a player, scale that difference by a constant
factor, and then add it to the player's initial rating. The only difference is
that rather than scaling by *K*, for a matchup with *N* players we scale by *K* 
times *N-1*. That is,

![equation](https://latex.codecogs.com/png.latex?R%27_A%20%3D%20R_A%20&plus;%20K%28N-1%29%28S_A%20-%20E_A%29)

The additional scaling by *N-1* isn't completely necessary if every matchup
always has the same number of players -- in that case we could simply use a
larger value of *K*. But in a scenario where some matchups have different
number of players, it is important to have larger Elo rating changes in 
larger matchups. For example, 1st place in a 10-player matchup should result in
a larger rating increase than 1st place in a 5-player matchup. Scaling by the
number of players accomplishes this.

CHANGELOG
# Change Log

## [0.2.0] -- 2021-04-21

### Added

* Estimate result probabilities with simulate_win_probabilities method in MultiElo class
* Improved logging, tests, documentation, and other small coding style improvements

### Changed

### Fixed

## [0.1.0] -- 2021-08-02

### Added

* Initial commit

### Changed

### Fixed



tests/test_elo.py
import pytest
from multielo import MultiElo
import numpy as np


def test_elo_defaults():
    # test if defaults are specified in config.py file
    try:
        from multielo.config import defaults
    except ImportError:
        assert False, "could not import defaults"

    # test that the expected keys are in defaults
    expected_keys = [
        "INITIAL_RATING",
        "K_VALUE",
        "D_VALUE",
        "SCORING_FUNCTION_BASE",
    ]
    for key in expected_keys:
        assert key in defaults, f"{key} is not included in defaults"
    for key in defaults:
        assert key in expected_keys, f"{key} is in defaults but is not one of the expected keys"


@pytest.mark.parametrize(
    "k, d, s, ratings, true_expected, true_new",
    [
        (32, 400, 1, np.array([1000, 1000]), [0.5, 0.5], [1016, 984]),
        (32, 400, 1, [1200, 1000], [0.75974693, 0.24025307], [1207.68809835, 992.31190165]),
        (32, 400, 1, [1000, 1200], [0.24025307, 0.75974693], [1024.31190165, 1175.68809835]),
        (32, 400, 1, np.array([1200, 800]), [0.90909091, 0.09090909], [1202.90909091, 797.09090909]),
        (64, 400, 1, [1200, 1000], [0.75974693, 0.24025307], [1215.37619669,  984.62380331]),
        (64, 800, 1, [1200, 1000], [0.640065, 0.359935], [1223.03584001,  976.96415999]),
        (32, 800, 1, [1200, 1000], [0.640065, 0.359935], [1211.51792001,  988.48207999]),
        (32, 200, 1, [1200, 1000], [0.90909091, 0.09090909], [1202.90909091,  997.09090909]),
        (32, 400, 1.5, [1200, 1000], [0.75974693, 0.24025307], [1207.68809835, 992.31190165]),
        (32, 400, 1, [1200, 1000, 900], [0.53625579, 0.29343936, 0.17030485],
         [1208.34629612, 1002.55321444, 889.10048944]),
        (32, 400, 1, [1000, 1200, 900], [0.29343936, 0.53625579, 0.17030485],
         [1023.88654777, 1187.01296279, 889.10048944]),
        (32, 400, 1.25, [1200, 1000, 900], [0.53625579, 0.29343936, 0.17030485],
         [1209.98732176, 1000.9121888, 889.10048944]),
        (32, 400, 1.5, [1200, 1000, 900], [0.53625579, 0.29343936, 0.17030485],
         [1211.39391517, 999.50559539, 889.10048944]),
        (32, 400, 2, [1200, 1000, 900], [0.53625579, 0.29343936, 0.17030485],
         [1213.67962945, 997.21988111, 889.10048944]),
        (32, 400, 1.25, [1200, 1000, 900, 1050], [0.38535873, 0.21814249, 0.13458826, 0.26191052],
         [1214.82857088, 1009.6423915, 900.67244749, 1024.85659012]),
    ]
)
def test_elo_changes(k, d, s, ratings, true_expected, true_new):
    """
    Test some known values to make sure Elo is calculating the correct updates.

    :param ratings:
    :param true_expected:
    :param true_new:
    :return:
    """
    elo = MultiElo(k_value=k, d_value=d, score_function_base=s)
    assert np.allclose(elo.get_expected_scores(ratings), true_expected)
    assert np.allclose(elo.get_new_ratings(ratings), true_new)


def test_zero_sum():
    # make sure expected scores sum to 1 and rating changes are zero sum
    for n_players in [2, 3, 4, 10]:
        for _ in range(10):
            k = np.random.uniform(16, 64)
            d = np.random.uniform(200, 800)
            ratings = np.random.uniform(600, 1400, size=n_players)
            elo = MultiElo(k_value=k, d_value=d)
            assert np.allclose(elo.get_expected_scores(ratings).sum(), 1), \
                f"expected ratings do not sum to 1 for k={k}, d={d}, ratings={ratings}"
            assert np.allclose(elo.get_new_ratings(ratings).sum(), ratings.sum()), \
                f"rating changes are not zero sum for k={k}, d={d}, ratings={ratings}"


tests/test_player_tracker.py
from multielo import Player, Tracker, MultiElo
from multielo.config import defaults
import pandas as pd
import numpy as np


def test_default_player():
    player = Player("test")
    assert player.rating == defaults["INITIAL_RATING"]


def test_player_with_history():
    player = Player("test", rating_history=[(1, 1000), (2, 1100)], date=3)
    assert player.rating == defaults["INITIAL_RATING"]
    assert player.get_rating_as_of_date(1) == 1000
    assert player.get_rating_as_of_date(2) == 1100


def test_update_rating():
    player = Player("test", rating=1000)
    player.update_rating(new_rating=1200, date=1)
    assert player.rating == 1200


def test_get_rating_as_of_date():
    player = Player("test", rating=900, date="2020-01-01")
    player.update_rating(new_rating=1100, date="2020-02-01")
    player.update_rating(new_rating=1200, date="2020-03-01")
    player.update_rating(new_rating=1300, date="2020-04-01")
    assert player.get_rating_as_of_date("2019-12-31") == defaults["INITIAL_RATING"]
    assert player.get_rating_as_of_date("2020-01-15") == 900
    assert player.get_rating_as_of_date("2020-02-15") == 1100
    assert player.get_rating_as_of_date("2020-03-15") == 1200
    assert player.rating == 1300


def test_player_equality():
    player_1 = Player("test1", rating=1200)
    player_2 = Player("test2", rating=1000)
    player_3 = Player("test3", rating=800)
    player_4 = Player("test4", rating=800)
    assert player_1 > player_2
    assert player_3 < player_2
    assert player_3 == player_4


def test_tracker():
    # test with some known results
    elo = MultiElo(k_value=32, d_value=400, score_function_base=1)
    tracker = Tracker(elo_rater=elo, initial_rating=1000)
    data = pd.DataFrame({
        "date": ["2020-03-29", "2020-04-05", "2020-04-12"],
        "1st": ["Homer", "Lisa", "Lisa"],
        "2nd": ["Marge", "Bart", "Marge"],
        "3rd": ["Bart", "Homer", "Homer"]
    })
    tracker.process_data(data)
    homer = tracker.retrieve_existing_player("Homer")
    marge = tracker.retrieve_existing_player("Marge")
    bart = tracker.retrieve_existing_player("Bart")
    lisa = tracker.retrieve_existing_player("Lisa")
    assert homer.count_games() == 3
    assert marge.count_games() == 2
    assert bart.count_games() == 2
    assert lisa.count_games() == 2
    assert np.mean([homer.rating, marge.rating, bart.rating, lisa.rating]) == 1000
    assert homer == 977.4832443375076
    assert marge == 1000.5940386640012
    assert bart == 980.6241719618897
    assert lisa == 1041.2985450366014


tests/test_score_functions.py
import numpy as np
from multielo.score_functions import linear_score_function, create_exponential_score_function


def test_2_player_score_function():
    # should always return [1, 0]
    linear_scores = linear_score_function(2)
    assert np.allclose(linear_scores, np.array([1, 0])), \
        f"linear score function does not return [1, 0] for 2-player matchup: {linear_scores}"

    for base in np.random.uniform(1, 3, 10).tolist():
        func = create_exponential_score_function(base)
        exp_scores = func(2)
        assert np.allclose(exp_scores, np.array([1, 0])), \
            f"exponential score function does not return [1, 0] for 2-player matchup for base={base}: {exp_scores}"


def test_linear_score_function():
    for n in range(3, 11):
        scores = linear_score_function(n)
        score_diffs = np.diff(scores)
        assert np.allclose(scores.sum(), 1), f"linear score function does not sum to 1 for n={n}: {scores}"
        assert np.allclose(scores.min(), 0), \
            f"linear score function does not have minimum score of 0 for n={n}: {scores}"
        assert np.all(score_diffs < 0), \
            f"linear score function is not monotonically decreasing for n={n}: {scores}"


def test_exponential_score_function():
    for base in np.random.uniform(1, 3, 10).tolist():
        for n in range(3, 11):
            func = create_exponential_score_function(base)
            scores = func(n)
            score_diffs = np.diff(scores)
            assert np.allclose(scores.sum(), 1), \
                f"exponential score function does not sum to 1 for base={base}, n={n}"
            assert np.allclose(scores.min(), 0), \
                f"exponential score function does not have minimum score of 0 for n={n}, base={base}: {scores}"
            assert np.all(score_diffs < 0), \
                f"exponential score function is not monotonically decreasing for base={base}, n={n}: {scores}"
            # differences should also be monotonically increasing (less negative) for exponential score function
            assert np.all(np.diff(score_diffs) > 0), \
                f"exponential score function diffs are not monotonically increasing for base={base}, n={n}: {scores}"


multielo/score_functions.py
import numpy as np
from numpy import ndarray


def linear_score_function(n: int) -> ndarray:
    """
    With the linear score function the "points" awarded scale linearly from first place
    through last place. For example, improving from 2nd to 1st place has the same sized
    benefit as improving from 5th to 4th place.

    :param n: number of players
    :return: array of the points to assign to each place (summing to 1)
    """
    return np.array([(n - p) / (n * (n - 1) / 2) for p in range(1, n + 1)])


def create_exponential_score_function(base: float):
    """
    With an exponential score function with base > 1, more points are awarded to the top
    finishers and the point distribution is flatter at the bottom. For example, improving
    from 2nd to 1st place is more valuable than improving from 5th place to 4th place. A
    larger base value means the scores will be more weighted towards the top finishers.

    :param base: base for teh exponential score function (> 1)
    :return: a function that takes parameter n for number of players and returns an array
    of the points to assign to each place (summing to 1)
    """
    return lambda n: _exponential_score_template(n, base)


def _exponential_score_template(n: int, base: float) -> ndarray:
    if base < 1:
        raise ValueError("base must be >= 1")
    if base == 1:
        return linear_score_function(n)  # it converges to this as base -> 1

    out = np.array([base ** (n - p) - 1 for p in range(1, n + 1)])
    return out / sum(out)


multielo/multielo.py
import numpy as np
from typing import Union, List
import logging

from .config import defaults
from .score_functions import create_exponential_score_function


class MultiElo:
    """
    Generalized Elo for multiplayer matchups (also simplifies to standard Elo for 1-vs-1 matchups).
    Does not allow ties.
    """

    def __init__(
        self,
        k_value: float = defaults["K_VALUE"],
        d_value: float = defaults["D_VALUE"],
        score_function_base: float = defaults["SCORING_FUNCTION_BASE"],
        custom_score_function=None,
        log_base: int = 10,
        logger: logging.Logger = None,
    ):
        """
        :param k_value: K parameter in Elo algorithm that determines how much ratings increase or decrease
        after each match
        :param d_value: D parameter in Elo algorithm that determines how much Elo difference affects win
        probability
        :param score_function_base: base value to use for scoring function; scores are approximately
        multiplied by this value as you improve from one place to the next (minimum allowed value is 1,
        which results in a linear scoring function)
        :param custom_score_function: a function that takes an integer input and returns a numpy array
        of monotonically decreasing values summing to 1
        :param log_base: base to use for logarithms throughout the Elo algorithm. Traditionally Elo
        uses base-10 logs
        :param logger: logger to use (optional)
        """
        self.k = k_value
        self.d = d_value
        self._score_func = custom_score_function or create_exponential_score_function(base=score_function_base)
        self._log_base = log_base
        self.logger = logger or logging.getLogger(__name__)

    def get_new_ratings(self, initial_ratings: Union[List[float], np.ndarray]) -> np.ndarray:
        """
        Update ratings based on results. Takes an array of ratings before the matchup and returns an array with
        the updated ratings. Provided array should be ordered by the actual results (first place finisher's
        initial rating first, second place next, and so on).

        Example usage:
        >>> elo = MultiElo()
        >>> elo.get_new_ratings([1200, 1000])
        array([1207.68809835,  992.31190165])
        >>> elo.get_new_ratings([1200, 1000, 1100, 900])
        array([1212.01868209, 1012.15595083, 1087.84404917,  887.98131791])

        :param initial_ratings: array of ratings (float values) in order of actual results
        :return: array of updated ratings (float values) in same order as input
        """
        if not isinstance(initial_ratings, np.ndarray):
            initial_ratings = np.array(initial_ratings)
        n = len(initial_ratings)  # number of players
        actual_scores = self.get_actual_scores(n)
        expected_scores = self.get_expected_scores(initial_ratings)
        scale_factor = self.k * (n - 1)
        self.logger.debug(f"scale factor: {scale_factor}")
        return initial_ratings + scale_factor * (actual_scores - expected_scores)

    def get_actual_scores(self, n: int) -> np.ndarray:
        """
        Return the scores to be awarded to the players based on the results.

        :param n: number of players in the matchup
        :return: array of length n of scores to be assigned to first place, second place, and so on
        """
        scores = self._score_func(n)
        self._validate_actual_scores(scores)
        self.logger.debug(f"calculated actual scores: {scores}")
        return scores

    @staticmethod
    def _validate_actual_scores(scores: np.ndarray):
        if not np.allclose(1, sum(scores)):
            raise ValueError("scoring function does not return scores summing to 1")
        if min(scores) != 0:
            raise ValueError("scoring function does not return minimum value of 0")
        if not np.all(np.diff(scores) < 0):
            raise ValueError("scoring function does not return monotonically decreasing values")

    def get_expected_scores(self, ratings: Union[List[float], np.ndarray]) -> np.ndarray:
        """
        Get the expected scores for all players given their ratings before the matchup.

        :param ratings: array of ratings for each player in a matchup
        :return: array of expected scores for all players
        """
        self.logger.debug(f"computing expected scores for {ratings}")
        if not isinstance(ratings, np.ndarray):
            ratings = np.array(ratings)
        if ratings.ndim > 1:
            raise ValueError(f"ratings should be 1-dimensional array (received {ratings.ndim})")

        # get all pairwise differences
        diff_mx = ratings - ratings[:, np.newaxis]
        self.logger.debug(f"diff_mx = \n{diff_mx}")

        # get individual contributions to expected score using logistic function
        logistic_mx = 1 / (1 + self._log_base ** (diff_mx / self.d))
        np.fill_diagonal(logistic_mx, 0)
        self.logger.debug(f"logistic_mx = \n{logistic_mx}")

        # get each expected score (sum individual contributions, then scale)
        expected_scores = logistic_mx.sum(axis=1)
        n = len(ratings)
        denom = n * (n - 1) / 2  # number of individual head-to-head matchups between n players
        expected_scores = expected_scores / denom

        # this should be guaranteed, but check to make sure
        if not np.allclose(1, sum(expected_scores)):
            raise ValueError("expected scores do not sum to 1")
        self.logger.debug(f"calculated expected scores: {expected_scores}")
        return expected_scores

    def simulate_win_probabilities(
            self,
            ratings: Union[List[float], np.ndarray],
            n_sim: int = int(1e5),
            seed: int = None,
    ) -> np.ndarray:
        """
        Estimate the probability of each player finishing in each possible
        place using a simulation. Returns a matrix where (i, j) values are the
        probability that player i finishes in place j.

        To simulate a game including players in the
        ratings array, we generate a score for each player using a Gumbel
        distribution. If a player has rating R, then that player's score is
        sampled from a Gumbel(R, D) distribution, where D is the Elo D
        parameter. Then we rank the players in descending order of their
        scores to determine first place, second place, ..., last place. We
        count the number of times each player finishes in each place and then
        divide by the number of simulations to calculate the proportions.

        We generate scores using Gumbel distributions because of the property:
        ~~    Gumbel(a_1, b) - Gumbel(a_2, b) ~ Logistic(a_1 - a_2, b)    ~~

        The Logistic(a_1 - a_2, b) distribution is the same distribution that
        describes the pairwise win probability if two payers have Elo ratings
        a_1 and a_2. In other words, a score sampled from Gumbel(a_1, b) will
        be greater than a score sampled from Gumbel(a_2, b) with the same
        probability that a player with Elo rating a_1 will beat a player with
        Elo rating a_2 in a 1-on-1 matchup.

        :param ratings: array of ratings of the players involved
        :param n_sim: number of simulations to run
        :param seed: (optional) seed for random number generation

        :return: matrix (a numpy array) where (i, j) values are the probability
        that player i finishes in place j
        """
        if seed is not None:
            np.random.seed(seed)

        # sort so we always get the same result for same distinct ratings, but
        # keep track of original order
        idx = np.argsort(ratings)
        ratings = sorted(ratings)

        # simulate n_sim scores for each player from Gumbel distributions
        n_players = len(ratings)
        n_sim = int(n_sim)
        scores = np.zeros((n_players, n_sim))
        self.logger.debug(f"simulating {n_sim:,} scores for each player")
        for i, rating in enumerate(ratings):
            scores[idx[i], :] = _gumbel_sample(
                loc=rating,
                scale=self.d,
                size=int(n_sim),
                base=self._log_base
            )
            self.logger.debug(f"finished sampling {n_sim:,} scores for player {i+1} of {n_players}")

        # use the scores to decide the order of finish (highest score wins) and
        # create matrix with proportion of times each player finishes in each place
        result_mx = self._convert_scores_to_result_proportions(scores)
        self.logger.debug(f"finished simulation")
        return result_mx

    @staticmethod
    def _convert_scores_to_result_proportions(scores: np.ndarray) -> np.ndarray:
        """
        Take an array of scores with one row per player and one column per
        simulation, and return a matrix with one row per player and one column
        per place. Each (row, col) value in the returned matrix is the count of
        times player "row" finished in place "col".
        """
        # sort scores from high to low for each simulation
        results = np.argsort(-scores, axis=0)

        # put it into a matrix where row = player, column = place, value = count
        # of times player finished in place
        n = scores.shape[0]
        count_mx = np.zeros((n, n))
        for i, x in enumerate(results):
            counts = np.bincount(x, minlength=n)
            count_mx[:, i] = counts

        proportion_mx = count_mx / scores.shape[1]
        return proportion_mx


def _gumbel_sample(
        loc: float,
        scale: float,
        size: int = 1,
        base: float = np.exp(1),
) -> np.ndarray:
    """
    Sample from a Gumbel distribution (optionally with a different log base).

    :param loc: location parameter for distribution
    :param scale: scale parameter for distribution (> 0)
    :param size: number of samples to draw
    :param base: base for logarithm (defaults to natural log)

    :return: sample(s) from Gumbel distribution
    """
    if scale <= 0:
        raise ValueError("scale parameter for Gumbel distribution must be > 0")
    p = np.random.rand(int(size))
    return loc - scale * _log(-_log(p, base=base), base=base)


def _log(x, base=np.exp(1)):
    return np.log(x) / np.log(base)
